First draft for logging and provenance. (#94)
OCR-D/spec: logging.md

# Conventions for LOGGING
[See issues](https://github.com/OCR-D/spec/issues/11)

> All logging may be stored in files. 

Reviewer: It should be made clear that this is a different spec for different target audience than e.g. the [docs on the CLI](https://ocr-d.github.io/cli), which states

> Data printed to `STDERR` and `STDOUT` is captured linewise and stored as log data.

Maybe an introductory section, explaining the purpose of the document, who should read it and which are the essential requirements to implement.

Author: I'll write something.

> If so, they must be given in Provenance as a result. 

Reviewer: Provenance capitalized sounds kind-of biblical :) Seriously though, I'm missing context here, I don't understand what `if` is related to, what `being given in provenance` means and what a `result` is here.

## Log Levels

> (Is there any definition we could link to)

Reviewer 2: ```suggestion
(Is there any definition we could link to)
```

> (Is there any definition we could link to)

Reviewer 2: Unfortunately I am not aware of any...[RFC 5424](https://tools.ietf.org/html/rfc5424) is not super verbose on this and the best explanation I could find is from [Stackoverflow](https://stackoverflow.com/a/5278006/1919465).

> A more detailed description will be found [here](https://stackoverflow.com/questions/2031163/when-to-use-the-different-log-levels/5278006#5278006)

Reviewer: Since this is actually pertinent to developers IMHO it should be added to https://ocr-d.github.io/cli#logging.

The `core` implementation doesn't distinguish between TRACE and DEBUG.

Author: True. Trace is only for completeness and should not be used in the workflow.

### TRACE / ALL
This is the most verbose logging level to trace the path of the algorithm.
E.g.: Start/End/Duration of a method, even loops may be logged. 
Note that logging within loops can dramatically affect performance.
### DEBUG
This level is used if parameters and or the status of an algorithm/method should be logged. 
### INFO
Log information about used settings on application level.
### WARN
This level is used to log information about missing/wrong configurations which may lead to errors.
### ERROR
Log all events that produce no or a wrong result. 

> It is useful to provide so much information that could provide information about the cause.

Reviewer 2: This line isn't really clear to me?

## Format

> The format of the logging output have to formatted like this:

Reviewer 2: ```suggestion
The format of the logging output has to be formatted like this:
```

> ## Format

Reviewer: Already defined in https://ocr-d.github.io/cli#logging

TIMESTAMP LEVEL LOGGERNAME - MESSAGE
***Example:***
08:03:40.017 WARN edu.kit.ocrd.MyTestClass - A warn message
:warning: 
Since only the time stamp is logged, the log should be written to files that change daily.

## METS
### File Group

> File group holding logging have to start with prefix "LOG-"

Reviewer 2: ```suggestion
File groups holding logging have to start with prefix "LOG-"
```

> Since only the time stamp is logged, the log should be written to files that change daily.

Reviewer: IIUC, you want logging to be the outcome of a job/task/process right (because of the bit about provenance at the top). That would not be a continuous runnign process where log rotation was important. Or do you want to concatenate the logging files of processes into a big file?

Author: This is a clumsy phrase. The output should never be collected in a large file. The statement should be that the output can always be assigned to a work step.

> File groups holding logging have to start with prefix "LOG-"

Reviewer: So in terms of BagIt, you see log data as part of the payload? Provenance metadata, parametrization etc. also part of the payload? How to map between provenance, log data, processes and files?

Author: At least the provenance should be part of the payload. The provenance refers to the input/output files (via ID?), which can then be used for assignment. All provenances must then be combined to form a graph.

Reviewer: > At least the provenance should be part of the payload.

Still don't understand why. Provenance is by definition metadata and not about the thing but about the process that made the thing. Why should this be different from logging data e.g.?

My main problem here is that METS consists (for our purposes!) of files representing pages in file groups representing different manifestations of these pages. The file groups are the output of processes but they don't reflect processes directly.

Provenance is based on processes which can have arbitrary input/output. With special filegroups that don't reflect pages anymore, this consistency breaks. If provenance was to be tracked in METS (which I would not advise), I see these options:

1) Have a file group `X-PROV` for every file group `X` which a `mets:file` for every `mets:file` in `X` that points to the URL of the process that generated that file
2) Add a custom attribute or (re-/ab-)use an existing attritute on the file itself
3) Have some sort of mapping structure that maps XML IDs to the URL of the process that created them

The last one would at least be general enough to also handle cases like "which process created the table of contents". But it will still be a convention not a standard, extend the size of METS forever and have little use for the institutions (who have their own digitization workflow engines which have their own naming schemes and conventions for provenance).

@cneud @wrznr

Author: Provenance is important because it is the only way to reproduce data, but reproducibility is an essential characteristic of scientific activity.
By (our) definition, all files contained in the bagit container must also be referenced in METS. As a result, the provenance must also be referenced in METS if it is part of the payload.
Optionally, we can define it as metadata that should be stored in a tag directory of the bagit container, but then the coupling is relatively loose.  (See comment below)
Also the bagit import/export to/from workspace will become more difficult.
